src/probnum/filtsmooth/gaussfiltsmooth/kalman.py

```diff
@@ -1,128 +1,213 @@
 
+    def smooth(self, filter_posterior, **kwargs):
+        """
+        Apply Gaussian smoothing to a set of filtered means and covariances.
 
-class _DiscDiscKalman(Kalman):
-    """
-    Provides predict() and update() methods for Kalman filtering and
-    smoothing on discrete-discrete state space models.
-    """
-
-    def __init__(self, dynamod, measmod, initrv):
-        """Checks that dynamod and measmod are linear and moves on."""
-        if not issubclass(type(dynamod), DiscreteGaussianLinearModel):
-            raise ValueError(
-                "ContinuousDiscreteKalman requires " "a linear dynamic model."
-            )
-        if not issubclass(type(measmod), DiscreteGaussianLinearModel):
-            raise ValueError(
-                "DiscreteDiscreteKalman requires " "a linear measurement model."
-            )
-        super().__init__(dynamod, measmod, initrv)
-
-    def predict(self, start, stop, randvar, **kwargs):
-        """Prediction step for discrete-discrete Kalman filtering."""
-        mean, covar = randvar.mean, randvar.cov
-        if np.isscalar(mean) and np.isscalar(covar):
-            mean, covar = mean * np.ones(1), covar * np.eye(1)
-        dynamat = self.dynamicmodel.dynamicsmatrix(start, **kwargs)
-        forcevec = self.dynamicmodel.forcevector(start, **kwargs)
-        diffmat = self.dynamicmodel.diffusionmatrix(start, **kwargs)
-        mpred = dynamat @ mean + forcevec
-        ccpred = covar @ dynamat.T
-        cpred = dynamat @ ccpred + diffmat
-        return Normal(mpred, cpred), {"crosscov": ccpred}
+        Parameters
+        ----------
+        filter_posterior : KalmanPosterior
+            Posterior distribution obtained after filtering
 
-    def update(self, time, randvar, data, **kwargs):
-        """Update step of discrete Kalman filtering"""
-        return _discrete_kalman_update(
-            time, randvar, data, self.measurementmodel, **kwargs
+        Returns
+        -------
+        KalmanPosterior
+            Posterior distribution of the smoothed output
+        """
+        rv_list = self.smooth_list(
+            filter_posterior, filter_posterior.locations, **kwargs
+        )
+        return KalmanPosterior(
+            filter_posterior.locations, rv_list, self, with_smoothing=True
         )
 
+    def smooth_list(self, rv_list, locations, final_rv=None, **kwargs):
+        """
+        Apply smoothing to a list of RVs with desired final random variable.
+
+        Specification of a final RV is useful to compute joint samples from a KalmanPosterior object,
+        because in this case, the final RV is a Dirac (over a sample from the final Normal RV)
+        and not a Normal RV.
+
+        Parameters
+        ----------
+        rv_list : _RandomVariableList or array_like
+            List of random variables to be smoothed.
+        locations : array_like
+            Locations of the random variables in rv_list.
+        final_rv : RandomVariable, optional.
```

Comment:
Good question. I guess so, but not sure. Lets come back to this after the sampling PR is merged?

Reply:
Im not sure I completely understood you, do you mean that you will change that in the future, but still in this PR, or that you would like to postpone this for another PR? In case the sampling PR is merged while this one is still open, I think this change should be included in this PR.

Reply:
if the sampling PR is merged first, then of course I will get back to this one here. I am not sure though if this will happen... :)